mm0-rs/src/elab/lisp/parser.rs

```diff
@@ -492,7 +492,16 @@ impl<'a> LispParser<'a> {
     let (sp, x, stack) = self.def_var(e)?;
     let ir = self.def_ir(sp, es, stack)?;
     if self.ctx.len() == 0 {
-      self.spans.insert(sp, ObjectKind::Global(x));
+      match x {
```

Comment:
Instead of doing this at parse time, where you might accidentally pick up false positives if someone decides to name a local `refine-extra-args`, how about doing it at evaluation time? [Here](https://github.com/digama0/mm0/blob/master/mm0-rs/src/elab/lisp/eval.rs#L1401-L1405) is where global definitions get added to the environment; the inner if is triggered if it is a definition and not just a redefinition, and you can also check `self.data[a].graveyard` to see if it is the first ever definition (as opposed to a definition, then an undefinition, then another definition). But for a syntax hint probably we just want all definitions/undefinitions to get the doc.

Also, the three cases in this match can be simplified using a `TryFrom<AtomID> for BuiltinCallback` impl.

Reply:
Hmm, I thought the "if self.ctx.len() == 0" check was meant to make sure that a definition was global and not local, but you're right, it seems not, since hovering over an inner definition still gives the doc. However, this seems to imply that the code before my changes, which adds the ObjectKind::Global in the parser, is also wrong, since it will give documentation for global definitions even when you redefine it locally. And in fact this does happen, you can test it like this:

```
do {
  --| Documentation for `hello`
  (def (hello) 1) -- has the documentation
  (hello) -- has the documentation
  (def (a-test)
    (def (hello) 2) -- incorrectly has the documentation
    (hello)) -- correctly does not have documentation
  (hello) -- has the documentation
};
```

Reply:
You are right, that check is wrong. Checking for an empty context excludes the possibility of any local scopes with variables but if you just have a plain `(begin)` then it won't catch that. Evaluation can tell the difference because when you are in a `(begin)` there is a continuation for it on the stack, so by peeking the stack we can tell if we are at global scope.

Also, I want to add the ability to use `match` at the top level to introduce multiple global definitions at once at some point, so I don't like the idea of tying things to the syntax of `(def)` as much as the `Def(x)` IR. I'll think some more about how to fix this.